VoodooI2CHID/VoodooI2CHIDDevice.cpp

```diff
@@ -148,36 +184,49 @@ IOReturn VoodooI2CHIDDevice::getHIDDescriptorAddress() {
 void VoodooI2CHIDDevice::getInputReport() {
     IOBufferMemoryDescriptor* buffer;
     IOReturn ret;
-    unsigned char* report = (unsigned char *)IOMalloc(hid_descriptor.wMaxInputLength);
-
-    api->readI2C(report, hid_descriptor.wMaxInputLength);
+    int return_size;
+    unsigned char* report;
     
-    int return_size = report[0] | report[1] << 8;
+    if (I2C_TRYLOCK() == false) {
+        IOLog("%s::%s Skipping a HID read interrupt while other thread read/write I2C\n", getName(), name);
+        goto exit;
+    }
+    report = getMallocI2CIntr(hid_descriptor.wMaxInputLength);
+    report[0] = report[1] = 0;
+    ret = api->readI2C(report, hid_descriptor.wMaxInputLength);
+    if (ret != kIOReturnSuccess) {
+        I2C_UNLOCK();
```

Comment:
As I said above, `goto exit;` sets `read_in_progress = false;`. However, we unlock the mutex before assigning value. This again could lead to undefined behavior.

I think we should call `I2C_UNLOCK();` after `read_in_progress = false;` in `exit`